### PR TITLE
Integration Plugin with Paperless integration

### PR DESCRIPTION
**Pull request by fuhrysteve**
Hi @sbs20,

After wrestling with a robust way to get data from scanservjs to paperless and reviewing [the discussion here](https://github.com/sbs20/scanservjs/issues/351) as well as some of the existing solutions involving copying files, I thought I'd take a stab at putting together an unobtrusive way to build integrations directly into scanservjs.

Please note that this is an INCOMPLETE and UNTESTED proof-of-concept / draft, but I thought I'd run it by you sooner rather than later, since I am sure you are sensitive to merging in functionality that is secondary to the core purpose of scanservjs.

Please let me know if you like this direction enough that you might actually be open to merging it someday (once it's working)

### CODE REVIEW

File: packages/server/config/config.default.js [hunk 1]
```diff
@@ -82,6 +82,27 @@ module.exports = {
...
+    //     token: 'xxxxxxxxxx'
+    //   }
+    // };
+
+    // config.plugins = {
+    //   paperless: {
+    //     hostname: 'paperless.mysite.com',
+    //     https: true,
+    //     always_send: false,  // whether to upload every scan automatically
+    //     show_upload_icon: true, // shows the "Send to Paperless" upload button on Files page
```
Comment (torwag) [inline packages/server/config/config.default.js:103]: port is part of the constructor but not listed here yet. Maybe the URL could also contain the port like paperless.mydomain.xxx:8080.

**Review by torwag — COMMENTED** [review]
Is there a particular reason why you went for a plugin integration vs. a pipeline. I planned something very much the same, but was thinking of adding it as a pipeline, so that I can select "Send2Paperless" in the Format section of ScanservJS.

File: packages/server/config/config.default.js [hunk 2]
```diff
@@ -82,6 +82,27 @@ module.exports = {
...
+
+
+    /* Integration Plugins */
+
+    // Paperless Integration example
+    //
+    // Prevent accessing secrets from the UI
+    // config.plugin_secrets = {
+    //   paperless: {
+    //     token: 'xxxxxxxxxx'
```
Comment (torwag) [inline packages/server/config/config.default.js:94]: A link or a short description, of how to obtain the token might be helpful (it is a rather unknown feature of paperless).

**Comment by sbs20** [comment]
Hi @fuhrysteve - thanks for taking the time to look at this.

So first off - I think you've identified a real gap in the software. As it stands today the only real workaround is to use a pipeline - which is what @torwag  is doing. The problem is that it's hijacking the creation process rather than firing an `onComplete` event. Your solution is better in this regard.

However, the reservation I have is that it starts requiring that integrations are built into the app itself or at the very least require some javascript coding. Once that happens, then this repo (namely me) is responsible for maintaining the integration and dealing with support issues, which is not something I have any appetite (or rather, time) for. Which is a shame, because the code you've written is very elegant.

My suggestion would be to have a `afterScan(FileInfo)` hook / callback / event whatever which fires when the scan finishes. The default implementation would be to do nothing. This would need to be wired into the Config load and `user-options.js`

At that point, it's entirely the responsibility of the implementor to write whatever code they want. The app could offer a few sample options and helpers.

`config.default.js`
```javascript
module.exports = {
  /**
   * @param {Configuration} config 
   */
  afterConfig(config) {
  },

  /**
   * @param {ScanDevice[]} devices 
   */
  afterDevices(devices) {
  }

  /**
   * This method is called after a scan completes. You can use this to ...
   * @param {FileInfo} fileInfo 
   */
  afterScan(fileInfo) {
    // insert whatever code you like here
    //  * you can import other javascript files
    //  * or just execute a command line
  }
};
```

We could create an integration class which offers a few helpers

`integration.js`
```javascript
class Integration {
  static async webhook(options) {
    // do HTTP request with fileInfo payload
    // to be implemented
  }

  static async command(cmdBuilder) {
    Process.spawn(cmdBuilder.build());
  }
}
```

And then tie those together:

`config.local.js`
```javascript
  afterScan(fileInfo) {
    // A user can add this
    Integration.command(new CmdBuilder('cp')
      .arg(fileInfo.fullname)
      .arg('/my/path/for/something'));

    // or you could paste your paperless.upload() here
  }
```

And we could document sample code for paperless. Thoughts?

**Comment by torwag** [comment]
I would love to see not only a defined way of calling an external function but also a way to signal scanservjs about the status.

Maybe a helper function
return_status(value, text) and the definition that certain return values will trigger different actions of scanservjs
(silently skip, log, show error pop-up, show success pop-up)... I always fear that some piece of such a toolchain brake and a non teachy user kills hours going on without noticing.

**Comment by fuhrysteve** [comment]
Lots of good feedback to go through here, but thought I'd take a stab at @sbs20's concern about making "plugin code" something that can be maintained outside of this project.

What if we made a dead-simple plugin system with a registry. The registry could simply be something like `plugin-registry.txt` and it would contain a list of known plugins for scanservjs. For example:

```
"John's Dropbox Integration 2.0",git@github.com:johndoe/scanservjs-dropbox-plugin.git@master
"Really Good Paperless Plugin",git@github.com:blah/some-paperless-plugin.git@master
```

The user interface would have a button that would let you install any plugin in the `plugin-registry.txt` file.

Any thoughts on that approach, @sbs20? I think the architecture could be largely the same as I've already thrown together, but with the caveat that any of the "paperless" code would be maintained fully in the plugin repository (by some other maintainer), and all scanservjs would need to maintain are the registry & a handful of plugin hooks / callbacks / vue.js components / whatever

**Comment by sbs20** [comment]
> I appreciate that, and thanks for being honest. I like the approach you suggested in the config file and would be happy to proceed in that direction if you're ok with it.

Are you saying you're up for developing the scaffolding to add the hook for `afterScan` and adding to `UserOptions` and all that? If so - that would be great 😄 . I'm also thinking about an `afterPage` hook too - but haven't really thought much beyond that.

> I would like to find a mechanism that would allow users to inject functionality into the UI, such as a button to send to some other system. If you have any ideas for how to best achieve that or if you think that's otherwise too invasive, please let me know!

The only thing I can think of is an optional config that the app knows about which will inject a script into `index.html`. At that point you can effectively do whatever you want with the DOM. That said, I haven't really thought about it at all beyond that and it may be a pain to write.

Out of interest, what sort of problem are you looking to solve? i.e. what button, where, to do what?

**Comment by fuhrysteve** [comment]
> Out of interest, what sort of problem are you looking to solve? i.e. what button, where, to do what?

Probably 2x possible configurable modes: (1) send everything to Paperless, or (2) a button on the Files page to send an individual file to Paperless.

I think it is really important to have an excellent easy-to-setup integration between Scanservjs & Paperless that you could setup in 5 minutes or less.  Scanservjs + Paperless really is a killer combo for folks who want to self-host apps at home. If there were an excellent easy to setup integration between the two projects, I think it would drive a lot more interest from would-be contributors to use & contribute to both projects.

**Comment by sbs20** [comment]
See #484

**Comment by fuhrysteve** [comment]
Closing this for now - going to check out that `afterScan` functionality soon!

**Comment by sbs20** [comment]
> Probably 2x possible configurable modes: (1) send everything to Paperless, or (2) Send one-at-a-time via a button on the Files page to send an individual file to Paperless.

I think this may be resolved shortly by #556 (Issue #504)

**Pull request closed without merging**